src/coreclr/vm/jitinterface.cpp

```diff
@@ -4446,6 +4446,55 @@ bool CEEInfo::isMoreSpecificType(
     return result;
 }
 
+/*********************************************************************/
+// Returns TypeCompareState::Must if cls is known to be an enum.
+// For enums with known exact type returns the underlying
+// type in underlyingType when the provided pointer is
+// non-NULL.
+// Returns TypeCompareState::May when a runtime check is required.
+TypeCompareState CEEInfo::isEnum(
+        CORINFO_CLASS_HANDLE        cls,
+        CORINFO_CLASS_HANDLE*       underlyingType)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    TypeCompareState result = TypeCompareState::May;
+
+    if (underlyingType != NULL)
+    {
+        *underlyingType = 0;
+    }
+
+    JIT_TO_EE_TRANSITION();
+
+    TypeHandle th(cls);
+
+    if (!th.IsNull() && !th.IsTypeDesc())
```

Comment:
`IsTypeDesc` is too conservative here. This should return `TypeCompareState::May` for `IsGenericVariable` only.

Reply:
AsMethodTable asserts for types like pointers without this check.